src/compiler/commandLineParser.ts

```diff
@@ -1843,8 +1843,10 @@ namespace ts {
             if (hasProperty(raw, "files") && !isNullOrUndefined(raw.files)) {
                 if (isArray(raw.files)) {
                     filesSpecs = <ReadonlyArray<string>>raw.files;
-                    if (filesSpecs.length === 0) {
-                        createCompilerDiagnosticOnlyIfJson(Diagnostics.The_files_list_in_config_file_0_is_empty, configFileName || "tsconfig.json");
+                    const hasReferences = hasProperty(raw, "references") && !isNullOrUndefined(raw.references);
+                    const hasZeroOrNoReferences = !hasReferences || raw.references.length === 0;
+                    if (filesSpecs.length === 0 && hasZeroOrNoReferences) {
+                        errors.push(createCompilerDiagnostic(Diagnostics.The_files_list_in_config_file_0_is_empty, configFileName || "tsconfig.json"));

```

Comment:
This is incorrect. You want to report this on files: [] in TsConfigFile if present otherwise createCompilerDiagnostic.,

Reply:
sorry I'm not understanding -- we check that `files` is a property of `raw` before going into this block:

```
if (hasProperty(raw, "files") && !isNullOrUndefined(raw.files))
```

My understanding is that `raw` will contain `files` in the case that `TsConfigFile` has that property.

Once we establish that `files` exists we make the assertion that the length is zero and that there are zero or no references. Am I missing something?

Reply:
Thanks for clarifying @sheetalkamat. In that case, I am thinking of making this check in `parseOwnConfigOfJsonSourceFile`, since that function will have access to the `valueNode` in `onSetValidOptionKeyValueInRoot`.

I could add an `emptyFilesValueNode` variable in the outer scope of `parseOwnConfigOfJsonSourceFile` and set that variable in `onSetValidOptionKeyValueInRoot` when the key === `"files"` and its length is `0`. Then, I can check the generated json from `convertToObjectWorker` to make an assertion along these lines:

```
if (emptyFilesValueNode && jsonHasZeroOrNoReferences) {
    errors.push(createDiagnosticForNodeInSourceFile(...));
}
```

If you have a better way of doing this please let me know. Otherwise I will make this change later today.

Reply:
No that's unnecessary. You can get the node from configFile by: 
```typescript
firstDefined(getTsConfigPropArray(tsConfigSourceFile, "files"), property => property.initializer)
```

Reply:
Much better thank you! I have updated code and unit tests to include location data when source file is present.